File: Kotlin/2. Queue/Circular_Queue.kt
interface MyQueue {
    fun enque(data: Any): Unit;
    fun deque(): Any;
    fun deQueOrNull(): Any?;
}

class ArrayQueue(private val capacity: Int) : MyQueue {
    private val array: Array<Any?> = arrayOfNulls<Any?>(capacity);

    private var _totalCount: Int = 0;
    val totalCount: Int
        get() = _totalCount

    private var front: Int = 0;
    private var rear: Int = 0;

    override fun enque(data: Any) {
        if (totalCount == capacity) throw Exception("Queue is Full !")
        if (rear == capacity) rear = 0;

        println("enque() : array[$rear]")   // 로깅
        array[rear++] = data;
        _totalCount++;
    }

    override fun deque(): Any {
        if (_totalCount == 0) throw Exception("Queue is Empty !")
        if (front == capacity) front = 0;

        println("deque() : array[$front]")  // 로깅
        val out: Any = array[front++]!!;
        _totalCount--;
        return out;
    }

    override fun deQueOrNull(): Any? {
        if (_totalCount == 0) return null;
        if (front == capacity) front = 0;

        val out: Any = array[front++]!!;
        _totalCount--;
        return out;
    }
}

class Order(
    private val id: Int,
    private val name: String,
) {}

fun main() {
    val orders = listOf(
        Order(1, "새우버거"),
        Order(1, "불고기버거"),
        Order(1, "치즈버거"),
    )
    val q = ArrayQueue(5);
    q.enque(orders[0]);
    q.enque(orders[1]);
    q.deque()
    q.deque()

    println(q.totalCount);
}

File: C/Queue/Circular_Queue.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "test.h"

QUEUE* Queue(int capacity) {
	QUEUE* q = (QUEUE*)calloc(1, sizeof(QUEUE));

	q->capacity = capacity;
	q->totalCount = 0;
	q->pArray =  calloc(capacity, sizeof(void*)); // 6이라면 6칸 배열 생성
	q->front = 0;
	q->rear = 0;
	return q;
}

void enque(QUEUE* q, void* data) {
	if (q->capacity == q->totalCount) {
		printf("용량 초과입니다\n");
		return;
	}
	q->pArray[q->rear] = data;
	printf("enque : array[%d] = %p\n", q->rear, data);
	q->totalCount++;
	if (++q->rear == q->capacity) {  // idx값 초과 시, 0으로 초기화
		q->rear = 0;
	}
}

void* deque(QUEUE* q) {
	if (q->totalCount == 0) {
		printf("추출할 요소가 없습니다\n");
		return;
	}

	void* out = q->pArray[q->front];
	printf("deque : array[%d] = %p\n", q->front, out);

	q->totalCount--;
	if (++q->front == q->capacity) {  // 더미 노드에 도달 시, 0으로 초기화
		q->front = 0;
	}
	return out;
}

int getCount(QUEUE* q) {
	return q->totalCount;
}

void destroyQueue(QUEUE* q) {
	free(q->pArray);
	free(q);
	return;
}

int main() {
	ORDER orders[] = {
		1, "새우버거",
		2, "불고기버거",
		3, "감자버거",
		4, "더블버거",
		5, "치즈버거",
	};

	QUEUE* q = Queue(5);

	enque(q, &orders[0]);
	enque(q, &orders[1]);
	deque(q);

	printf("total : %d\n", q->totalCount);
	destroyQueue(q);
}